logexp/cmd/main.go

```diff
@@ -0,0 +1,257 @@
+		isMaster = true
+		logfiles = append(logfiles, masterLogs...)
+	} else {
+		glog.Infof("VM identified as kubernetes node")
+		isMaster = false
+		logfiles = append(logfiles, nodeLogs...)
+	}
+
+	// Select cloud-provider specific logs.
+	switch kubernetesProvider {
+	case "gce", "gke":
+		logfiles = append(logfiles, gceLogs...)
+	case "kubemark":
+		logfiles = append(logfiles, gceLogs...)  // TODO(shyamjvs): Make this variable wrt kubemark's real provider.
+		if ! isMaster && enableHollowNodeLogs {
+			logfiles = append(logfiles, kubemarkLogs...)
+		}
+	case "aws":
+		logfiles = append(logfiles, awsLogs...)
+	default:
+		glog.Warningf("Unknown cloud provider '%v' provided, skipping any provider specific logs", kubernetesProvider)
+	}
+
+	// Select system/service specific logs.
+	if _, err := os.Stat("/workspace/etc/systemd/journald.conf"); err == nil {
+		glog.Infof("Journalctl found on host. Collecting systemd logs")
+		createSystemdLogfiles(logDir, isMaster)
+	} else {
+		glog.Infof("Journalctl not found on host (%v). Collecting supervisord logs instead", err)
+		logfiles = append(logfiles, kernelLog)
+		logfiles = append(logfiles, initdLogs...)
+		logfiles = append(logfiles, supervisordLogs...)
+	}
+
+	// Copy all the logfiles that exist, to logDir.
+	for _, logfile := range logfiles {
+		logfileFullPath := filepath.Join(localLogPath, logfile + ".log*")  // Append .log* to copy rotated logs too.
+		cmd := exec.Command("/bin/sh", "-c", fmt.Sprintf("cp %v %v", logfileFullPath, logDir))
+		if _, err := cmd.Output(); err != nil {
+			glog.Warningf("Failed to copy any logfiles with pattern '%v': %v", logfileFullPath, err)
+		}
+	}
+}
+
+// Effeciently upload all logfiles inside logDir to GCS.
+func uploadLogfilesToGCS(logDir string) error {
+	cmd := exec.Command("/bin/sh", "-c", fmt.Sprintf("ls %v/*", logDir))
+	if output, err := cmd.Output(); err != nil {
+		glog.Errorf("%v", )
+		return fmt.Errorf("Could not list any logfiles: %v", err)
+	} else {
+		glog.Infof("List of logfiles available: %v", output)
+	}
+
+	// Upload the files with compression and parallelism (speeds up on good network)
+  	// along with setting their ACL to make them publicly readable.
+  	gcsLogPath = filepath.Join(gcsArtifactsPath, nodeName)
+  	if ! strings.HasPrefix(gcsLogPath, "gs://") {
+  		// Hack to counter bash's simplification on multiple slashes.
+  		gcsLogPath = strings.Replace(gcsLogPath, "gs:/", "gs://", 1)
+  	}
+  	glog.Infof("Uploading logfiles to GCS at path '%v'", gcsLogPath)
+  	var uploadAttempt int
+  	for uploadAttempt = 0; uploadAttempt < 3; uploadAttempt++ {
+  		cmd := exec.Command("gsutil", "-m", "-q", "-o", "GSUtil:use_magicfile=True",
+  		 					"cp", "-a", "public-read", "-c", "-z", "log,txt,xml",
+  		 					logDir + "/*", gcsLogPath)
+  		if _, err := cmd.Output(); err != nil {
+  			glog.Errorf("Attempt %v to upload to GCS failed: %v", uploadAttempt, err)
+  			continue
+  		}
+  		break
+  	}
+  	if uploadAttempt == 3 {
```

Comment:
Included the last one in the returned error (as technically the final attempt failing is what should be considered the error). Is it fine?